akka-cluster/src/main/scala/akka/cluster/Cluster.scala

```diff
@@ -313,17 +317,45 @@ class Cluster(val system: ExtendedActorSystem) extends Extension {
    * to defer some action, such as starting actors, until the cluster has reached
    * a certain size.
    */
+  @throws(classOf[TerminatedClusterException])
   def registerOnMemberUp[T](code: ⇒ T): Unit =
-    registerOnMemberUp(new Runnable { def run = code })
+    registerOnMemberUp(new Runnable { def run() = code })
 
   /**
    * Java API: The supplied callback will be run, once, when current cluster member is `Up`.
    * Typically used together with configuration option `akka.cluster.min-nr-of-members'
    * to defer some action, such as starting actors, until the cluster has reached
    * a certain size.
+   * @throws TerminatedClusterException(String:message) Throws when the cluster have been shutdown.
    */
-  def registerOnMemberUp(callback: Runnable): Unit = clusterDaemons ! InternalClusterAction.AddOnMemberUpListener(callback)
+  @throws(classOf[TerminatedClusterException])
+  def registerOnMemberUp(callback: Runnable): Unit = {
+    if (_isTerminated.get())
+      throw new TerminatedClusterException("Cluster have been shutdown,you must restart the actor system to rejoin the cluster.")
+    else
+      clusterDaemons ! InternalClusterAction.AddOnMemberUpListener(callback)
+  }
 
+  /**
+   * The supplied thunk will be run, once, when current cluster member is `Removed`.
+   * Typically used together `cluster.leave(cluster.selfAddress)` and then `system.shutdown()`.
+   */
+  @throws(classOf[TerminatedClusterException])
+  def registerOnMemberRemoved[T](code: ⇒ T): Unit =
+    registerOnMemberRemoved(new Runnable { override def run(): Unit = code })
+
+  /**
+   * Java API: The supplied thunk will be run, once, when current cluster member is `Removed`.
+   * Typically used together `cluster.leave(cluster.selfAddress)` and then `system.shutdown()`.
+   * @throws TerminatedClusterException(String:message) Throws when the cluster have been shutdown.
+   */
+  @throws(classOf[TerminatedClusterException])
+  def registerOnMemberRemoved(callback: Runnable): Unit = {
+    if (_isTerminated.get())
+      throw new TerminatedClusterException("Cluster have been shutdown.")
```

Comment:
Yes, running this one in the current thread and the other one not at all sounds like a sensible thing to do; please also document that clearly so people are not surprised. The rationale is that once the cluster has started up, user code might wish to install some cleanup handling, but the cluster might already be shutting down, and depending on the race is not healthy.